<h1>
    <img src="https://github.com/cribbslab/phylotres/blob/main/phylotres/util/phylotres_sign.png?raw=true" width="300" height="100">
    <br>
</h1>

![](https://img.shields.io/pypi/v/phylotres?logo=PyPI)
![](https://img.shields.io/badge/last_released-Jul14._2023-green.svg)
![](https://img.shields.io/badge/phylotres-executable-519dd9.svg)


<!-- ![Build](https://github.com/2003100127/phylotres/actions/workflows/build.yml/badge.svg) -->

###### tags: `computational biology`, `sequencing read simulation`

## Overview

```angular2html

```

PhyloTres is a Python toolkit for simulating sequencing reads at the single-locus, bulk RNA-seq, and single-cell levels. It is implemented based on phylogenetic tree-based methods, which allows for ultra-fast simulation read generation. PhyloTres allows both short-read and long-read sequencing read simulation, and substitution and indel (insertion and deletion) errors added to reads. PhyloTres implements a very flexible read generation framework, which allows users to design their simulated reads in any forms and structures. PhyloTres can vastly help both computational and experimental researchers to swiftly test their sequencing method ideas.

## Documentation

Website: xxx

## Installation

### Using pip (recommended)

```sh
# create a conda environment
conda create --name phylotres python=3.11

# activate the conda environment
conda activate phylotres

# the latest version
pip install phylotres --upgrade
```

## Citation

Please cite our work if you use PhyloTres in your research.

## Contact

Please report any questions on [issue](https://github.com/2003100127/phylotres/issues) pages.
