
`tresor.sc.simu_ampl_rate` is a Python function in charge of simulating reads with respect to a series of amplication rates at the scRNA-seq level.

# Usage

We take the following command as an example to generate FastQ files at amplication efficiencies from 0.1 to 1.

=== "Python"

    ``` py hl_lines="27-31 34-36"
    import tresor as ts

    for perm_i in range(1):
        print(perm_i)
        ts.sc.simu_ampl_rate(
            # initial sequence generation
            gmat=gmat,

            len_params={
                'umi': {
                    'umi_unit_pattern': 3,
                    'umi_unit_len': 12,
                },
                'seq': 100,
            },
            material_params={
                'fasta_cdna_fpn': to('data/Homo_sapiens.GRCh38.cdna.all.fa.gz'),  # None False
            },
            seq_num=50,
            working_dir=to('data/simu/docs/'),
    
            condis=['umi', 'seq'],
            sim_thres=3,
            permutation=0,
    
            # PCR amplification
            ampl_rates=np.linspace(0.1, 1, 10),
            err_route='sptree',  # bftree sptree err1d err2d mutation_table_minimum mutation_table_complete
            pcr_error=1e-4,
            pcr_num=10,
            err_num_met='nbinomial',
            
            # PCR amplification
            seq_error=0.01,
            # seq_sub_spl_number=200, # None
            seq_sub_spl_rate=0.333,

            use_seed=True,
            seed=1,
    
            verbose=False,  # True False
            mode='short_read',  # long_read short_read
    
            sv_fastq_fp=to('data/simu/docs/'),
        )
    ```

=== "Shell"

    ``` c++ linenums="1"
    tresor amplrate_sc \
    -cfpn ./tresor/data/amplrate_sl.yml \
    -snum 50 \
    -permut 0 \
    -sthres 3 \
    -wd ./tresor/data/simu/ \
    -md short_read \
    -is True \
    -vb True
    ```


# Attributes
!!! Illustration

    === "Python"
        | Attribute      | Description                          |
        | :---------- | :----------------------------------- |
        | `seq_num` | number of RNA molecules. `50` by default  |
        | `len_params` | lengths of different components of a read |
        | `seq_params` | sequences of different components of a read, It allows users to add their customised sequences |
        | `material_params` | a Python dictionary. Showing if cDNA libraries are provided, please use key word `fasta_cdna_fpn`. The human cDNA library can be downloaded through the [Ensembl genome](https://ftp.ensembl.org/pub/release-112/fasta/homo_sapiens/cdna) database |
        | `ampl_rates` | list of float numbers ranging from 0 to 1 |
        | `err_route` | the computational algorithm to generate errors. There are 6 methods, including `bftree`, `sptree`, `err1d`, `err2d`, `mutation_table_minimum`, and `mutation_table_complete`. |
        | `pcr_error` | PCR error rate |
        | `pcr_num` | number of PCR cycles to amplify reads |
        | `err_num_met` | the method to generate errors, that is, `binomial` or `nbinomial` |
        | `seq_error` | sequencing error rate |
        | `seq_sub_spl_number` | number of subsampling PCR amplified reads. It exists when `seq_sub_spl_rate` is specified to `None` |
        | `seq_sub_spl_rate` | rate of subsampling PCR amplified reads. It exists when `seq_sub_spl_number` is specified to `None` |
        | `sv_fastq_fp` | folder to save FastQ files |
        | `is_seed` | if seeds are used to simulate sequencing libraries. This is designed to make in silico experiments reproducible |
        | `working_dir` | working directory where all simulation results are about to be saved |
        | `condis` | names of components that a read contains. It can contains an unlimited number of read components |
        | `sim_thres` | similarity threshold. `3` by default |
        | `permutation` | permutation times |
        | `mode` | `long_read` or `short_read` |
        | `verbose` | whether to print intermediate results |
        
    === "Shell"
        | Attribute      | Description                          |
        | :---------- | :----------------------------------- |
        | `cfpn` | location to the yaml configuration file. Users can specify the atrributes illustrated on the **Python** tab in the `.yml` file. |
        | `snum` | number of sequencing molecules |
        | `permut` | permutation times |
        | `sthres` | similarity threshold. `3` by default |
        | `wd` | working directory where all simulation results are about to be saved |
        | `md` | `long_read` or `short_read` mode |
        | `is` | if seeds are used to simulate sequencing libraries. This is designed for reproducible in silico experiments |
        | `vb` | whether to print intermediate results |


# Output
## Console
``` py
======>simulation completes in 0.02896857261657715s
======>simulation completes in 0.08196878433227539s
======>simulation completes in 0.20600128173828125s
======>simulation completes in 0.4779694080352783s
======>simulation completes in 1.5829687118530273s
======>simulation completes in 3.1749932765960693s
======>simulation completes in 5.8459696769714355s
======>simulation completes in 10.428000450134277s
======>simulation completes in 18.868970155715942s
======>simulation completes in 33.60697078704834s
Finished!
```

## Understanding files
The resultant files of the simulated reads are shown as follows.

<figure markdown="span">
  ![Image title](../../../img/simu/ampl_rate/sc.jpg){ width="800" }
  <figcaption><strong>Fig</strong> 1. Generated FastQ files</figcaption>
</figure>
